feat: add Gordon surface implementation and test

gongfan99 opened the pull request:
modified:   pyproject.toml
	modified:   src/build123d/topology/two_d.py
	modified:   tests/test_direct_api/test_face.py

gongfan99 committed:
feat: add Gordon surface implementation and test
	modified:   pyproject.toml
	modified:   src/build123d/topology/two_d.py
	modified:   tests/test_direct_api/test_face.py

gumyr commented on src/build123d/topology/two_d.py:
```diff
@@ -864,6 +865,38 @@ def extrude(cls, obj: Edge, direction: VectorLike) -> Face:
         return Face(TopoDS.Face_s(_extrude_topods_shape(obj.wrapped, direction)))
 
+    @classmethod
+    def gordon_surface(
```
Let's call this `make_gordon_surface` to align with the others like `make_bezier_surface`.

gongfan99 replied:
The function name has been changed to `make_gordon_surface`

gumyr commented on src/build123d/topology/two_d.py:
```diff
@@ -864,6 +865,38 @@ def extrude(cls, obj: Edge, direction: VectorLike) -> Face:
+        Returns:
+            Face: the interpolated Gordon surface
+        """
+        ocp_profiles = [BRep_Tool.Curve_s(edge.wrapped, 0, 1) for edge in profiles]
```
There will be a typing error here when one uses the OCP headers.  I've been slowly eliminating them to get a clean `mypy` run even with OCP headers present (they aren't present when `mypy` is run in the GitHub actions).  To "fix" these I do (from `Mixin1D.is_closed`):
```py
        if self.wrapped is None:
            raise ValueError("Can't determine if empty Edge or Wire is closed")
```
I'm uncertain of the real value of this type of error message to the user but a clean `mypy` run certainly has value when developing new stuff.  I get 54 typing problems with this file before this change so there is work to be done here.

gongfan99 replied:
I didn't know the type error can be fixed this way. Thanks for the example. I will fix it.

gongfan99 replied:
The type error has been fixed per the given example.

gumyr reviewed (changes requested):
Very nice.  I'm happy to have a Gordon Surface as part of build123d.  Thank you for doing this.

jdegenstein commented on tests/test_direct_api/test_face.py:
```diff
@@ -502,6 +502,78 @@ def test_make_surface_patch_error_checking(self):
                 ]
             )
 
+    def test_gordon_surface(self):
```
Also per the renaming recommendation from Roger above, please rename the test accordingly (just makes it easier to find/filter later).

gongfan99 replied:
The test is renamed to `test_make_gordon_surface`

gumyr commented:
I'm confused about the type of the parameters to `interpolate_curve_network`.  
It looks like you are using this:
```py
def interpolate_curve_network(
    ucurves: List[Union[Geom_Curve, Geom_BSplineCurve]],
    vcurves: List[Union[Geom_Curve, Geom_BSplineCurve]],
    tolerance: float = 3e-4
) -> Geom_BSplineSurface:
```
Which as far as I can determine only take OCCT objects yet in the testing you pass `Spline` objects.  Where am I going wrong with this?  It would be very helpful if the functions exposed to build123d had a type API.

gongfan99 commented:
@gumyr Spline object has never directly been passed to interpolate_curve_network()
Spline is passed to Face.make_gordon_surface which converts it to OCP's Geom_Curve object to feed to interpolate_curve_network()

gumyr commented:
> @gumyr Spline object has never directly been passed to interpolate_curve_network()
> 
> Spline is passed to Face.make_gordon_surface which converts it to OCP's Geom_Curve object to feed to interpolate_curve_network()

Of course you are right. It still would be useful if the API was typed so the arguments could be statically checked.

gongfan99 commented:
@gumyr Could you elaborate what you mean by "API should be typed"?

The interpolate_curve_network() function is already typed as below:
```python
def interpolate_curve_network(
    ucurves: List[Union[Geom_Curve, Geom_BSplineCurve]],
    vcurves: List[Union[Geom_Curve, Geom_BSplineCurve]],
    tolerance: float = 3e-4
) -> Geom_BSplineSurface:
```

And interpolate_curve_network() function is the only interface of ocp_gordon package.

gumyr commented:
Yes you've typed the function but I'm getting these typing failures:
```
Run mypy --config-file mypy.ini src/build123d
  mypy --config-file mypy.ini src/build123d
  shell: /usr/bin/bash -e {0}
  env:
    UV_PYTHON: 3.10
    VIRTUAL_ENV: /home/runner/work/build123d/build123d/.venv
    UV_CACHE_DIR: /home/runner/work/_temp/setup-uv-cache
src/build123d/topology/two_d.py:108: error: Skipping analyzing "ocp_gordon": module is installed, but missing library stubs or py.typed marker  [import-untyped]
src/build123d/topology/two_d.py:108: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
Found 1 error in 1 file (checked 37 source files)
Error: Process completed with exit code 1.
```
I not 100% sure but I think the problem is a missing: src_py/ocp_gordon/py.typed (an empty file) which will indicate that your API is typed.

gongfan99 commented:
@gumyr I got it now. I will add an empty py.typed file.

gongfan99 committed:
fix: change function name to make_gordon_surface
fix: change the test name accordingly
fix: corrected the type error for Edge.wrapped
fix: change min version of ocp_gordon to 0.1.12
	modified:   pyproject.toml
	modified:   src/build123d/topology/two_d.py
	modified:   tests/test_direct_api/test_face.py

gongfan99 committed:
fix: add test for ValueError for gordon surface
	modified:   tests/test_direct_api/test_face.py

gongfan99 committed:
fix: minor adjust to test_make_gordon_surface
	modified:   tests/test_direct_api/test_face.py

gumyr commented:
Thank you!  This is an important bit of new functionality for build123d.  I think I'll try to update my Spitfire wing to use a couple of Gordon surfaces.

gumyr commented:
@gongfan99 are you on the build123d discord?

gongfan99 commented:
@gumyr just joined. It is much busier there than Reddit. Thanks for the nice build123d project.

mnesarco commented:
@gongfan99 This is a huge feature. Thank you.

gumyr commented:
@gongfan99 I'm trying to create an interesting example so I can show the community but having trouble.  I've raised an issue on your repo, I hope you don't mind.

gongfan99 commented:
@mnesarco Thank you! I am also a fan of your FreeCAD workbench (guitar design).
